Kudu.Core/Deployment/StructuredTextLogger.cs

```diff
@@ -0,0 +1,120 @@
+        private const string LogEntryPattern = "message=\"(.*?)\"\\s+time=\"(.*?)\"\\s+id=\"(.*?)\"\\s+type=\"(.*?)\"";
+        private readonly Regex LogEntryRegex = new Regex(LogEntryPattern, RegexOptions.IgnoreCase | RegexOptions.Singleline);
+        private readonly StructuredTextDocument<LogEntry> _structuredTextDocment;
+
+        private static IEnumerable<KeyValuePair<string, string>> EscapeChars = new Dictionary<string, string>
+        {
+            { "\"", "&quot;" }
+        }
+        .Union(StructuredTextDocument.NotAllowedSequances);
+
+        public StructuredTextLogger(string path, IAnalytics analytics, int depth = 0)
+        {
+            _depth = depth;
+            _path = path;
+            _analytics = analytics;
+            _structuredTextDocment = new StructuredTextDocument<LogEntry>(path,
+                e => string.Format(LogEntryTemplate, e.Message, e.LogTime, e.Id, (int)e.Type),
+                str =>
+                {
+                    var match = LogEntryRegex.Match(str);
+                    if (match.Success)
+                    {
+                        var message = UnsanitizeValue(match.Groups[1].Value);
+                        var time = DateTime.Parse(match.Groups[2].Value);
+                        var id = match.Groups[3].Value;
+                        var type = (LogEntryType)Int32.Parse(match.Groups[4].Value);
+                        return new LogEntry(time, id, message, type);
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format("the log line \"{0}\" is in an invalid format", str));
+                    }
+                });
+        }
+
+        public ILogger Log(string value, LogEntryType type)
+        {
+            try
+            {
+                value = SanitizeValue(value);
+                _structuredTextDocment.Write(new LogEntry(DateTime.UtcNow, Guid.NewGuid().ToString(), value, type), _depth);
+                return new StructuredTextLogger(_path, _analytics, _depth + 1);
+            }
+            catch(Exception e)
+            {
+                _analytics.UnexpectedException(e);
+                throw;
+            }
+        }
+
+        public IEnumerable<LogEntry> GetLogEntries()
+        {
+            try
+            {
+                return _structuredTextDocment.GetDocument().Select(e => e.LogEntry);
+            }
+            catch(Exception e)
+            {
+                _analytics.UnexpectedException(e);
+                throw;
+            }
+        }
+
+        public IEnumerable<LogEntry> GetLogEntryDetails(string entryId)
+        {
+            try
+            {
+                return _structuredTextDocment.GetDocument()
+                    .Where(s => s.LogEntry.Id.Equals(entryId, StringComparison.OrdinalIgnoreCase))
+                    .Select(e => e.Children.Select(c => c.LogEntry))
+                    .SelectMany(i => i);
```

Comment:
Seems a bit more generic than needed, since you can never have a case where multiple top level entries share the same id. Technically, it's more correct to find the one and only entry, and then just return its children.